nixos/modules/services/networking/pihole-ftl.nix

```diff
@@ -0,0 +1,330 @@
+          let
+            concatMapLines = f: list: builtins.concatStringsSep "\n    " (map f list);
+            escapeSQL = builtins.replaceStrings [ "'" ] [ "''" ];
+          in
+          ''
+            set -euo pipefail
+            pihole=${cfg.pihole}/bin/pihole
+
+            # If the database doesn't exist, it needs to be created with gravity.sh
+            if [ ! -f ${cfg.stateDirectory}/gravity.db ]; then
+              DONT_RESTART_FTL=1 $pihole -g
+            fi
+
+            ${concatMapLines
+              (list: ''$pihole -a adlist add \
+                ${strings.escapeShellArg (escapeSQL list.url)} \
+                ${strings.escapeShellArg (escapeSQL list.comment)}'')
+              cfg.adlists}
+
+            # Run gravity.sh to load any new adlists
+            DONT_RESTART_FTL=1 $pihole -g
+          '';
+      };
+
+      pihole-ftl = {
+        description = "Pi-hole FTL";
+
+        after = [ "network.target" "pihole-ftl-setup.service" ];
+        requires = [ "pihole-ftl-setup.service" ];
+        wantedBy = [ "multi-user.target" ];
+
+        serviceConfig = {
+          Type = "simple";
+          User = cfg.user;
+          Group = cfg.group;
+          AmbientCapabilities = [
+            "CAP_NET_BIND_SERVICE"
+            "CAP_NET_RAW"
+            "CAP_NET_ADMIN"
+            "CAP_SYS_NICE"
+          ];
+          ExecStart = "${cfg.package}/bin/pihole-FTL no-daemon -- --conf-file=${validatedDnsmasqConf}";
+          Restart = "on-failure";
+          # Hardening
+          NoNewPrivileges = true;
+          PrivateTmp = true;
+          PrivateDevices = true;
+          DevicePolicy = "closed";
+          ProtectSystem = "strict";
+          ProtectHome = "read-only";
+          ProtectControlGroups = true;
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          ReadWritePaths = [ cfg.stateDirectory logFile debugFile ];
+          RestrictAddressFamilies = "AF_UNIX AF_INET AF_INET6 AF_NETLINK";
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          MemoryDenyWriteExecute = true;
+          LockPersonality = true;
+        };
+      };
+    };
+
+    services.dnsmasq.settings.dhcp-leasefile = "${cfg.stateDirectory}/dhcp.leases";
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedUDPPorts = [ 53 ];
+      allowedTCPPorts = [ 53 ];
+    };
+
+    users.users.${cfg.user} = {
+      group = cfg.group;
+      isSystemUser = true;
+    };
+
+    users.groups.${cfg.group} = { };
+
+    environment.etc."pihole/setupVars.conf" = {
+      source = setupVarsConf;
+      user = cfg.user;
+      group = cfg.group;
+      mode = "444";
+    };
+
+    environment.etc."pihole/pihole-FTL.conf" = {
+      source = ftlConf;
+      user = cfg.user;
+      group = cfg.group;
+      mode = "444";
+    };
+
+    environment.systemPackages = [ pihole-script ];
+
+    services.logrotate.settings.pihole-ftl = {
```

Comment:
pihole also has a mechanism for pruning queries from the database: https://github.com/pi-hole/pi-hole/blob/master/advanced/Scripts/piholeLogFlush.sh

It may be worth adding something similar to the module

Reply:
we definitely want this otherwise the db grows into infinity